### PR TITLE
Improve display of view-file macro #239

### PR DESCRIPTION
**Pull request by ChiuchiuSorin**
Created modals and custom box macro to better display the files used with the `view-files` macro.

* Macro view:
macro
macro-select

* Full view:
full-view1
full-view2

* Table test:
table-test

* Thumbnail demo:
https://github.com/xwikisas/xwiki-pro-macros/assets/104432589/a1970431-2657-4600-acff-23733bcab5ba

  * The thumbnail option only returns a hyperlink to a modal to view the selected file, without an image preview, as currently there is no support for creating an image from an office or pdf file.

### CODE REVIEW

File: xwiki-pro-macros-ui/src/main/resources/Confluence/Macros/ViewFile.xml [hunk 1]
```diff
@@ -263,24 +386,140 @@
...
+        #customBox($width $height $escapedFilename)
+      #elseif($doc.getAttachment($unescapedFilename))
+        #set ($downloadUrl = $xwiki.getAttachmentURL($doc.getFullName(), $unescapedFilename))
+        &lt;div&gt;
+          &lt;a href="$downloadUrl" download&gt;$escapedFilename&lt;br&gt;&lt;/a&gt;
+        &lt;/div&gt;
+      #end
+    #elseif ($display == 'THUMBNAIL')
+      #if (($extension == 'pdf' &amp;&amp; $hasPDFViewer) || $officeExtensions.contains($extension))
+        #set ($modalId = $idName + 'DisplayContentModal')
```
Comment (trrenty) [inline xwiki-pro-macros-ui/src/main/resources/Confluence/Macros/ViewFile.xml:506]: You should escape the modal id in all the places that u use it to generate html.

File: xwiki-pro-macros-ui/src/main/resources/Confluence/Macros/ViewFile.xml [hunk 2]
```diff
@@ -263,24 +386,140 @@
...
+  &lt;/div&gt;
+#end
+#macro(customPresentationBox $width $height $escapedFilename)
+  #if ($width &amp;&amp; !$width.contains("%"))
+    #set($width = $width + "px")
+  #end
+  #if ($height &amp;&amp; !$height.contains("%"))
+    #set($height = $height + "px")
+  #end
+  &lt;div style="width: $width; height: $height; overflow: auto;"&gt;
```
Comment (trrenty) [inline xwiki-pro-macros-ui/src/main/resources/Confluence/Macros/ViewFile.xml:409]: Escape

File: xwiki-pro-macros-ui/src/main/resources/Confluence/Macros/ViewFile.xml [hunk 3]
```diff
@@ -263,24 +386,140 @@
...
+#end
+#macro (displayContent $escapedFilename, $modalId)
+  &lt;div class="modal fade" id="$modalId" tabindex="-1" role="dialog"
+      aria-labelledby="${modalId}Label"&gt;
+    &lt;div class="modal-dialog" style="max-height: 90vh; width: 80%;" role="document"&gt;
+      &lt;div class="modal-content"&gt;
+        &lt;div class="modal-header"&gt;
+          &lt;button type="button" class="close" data-dismiss="modal" aria-label="Close"&gt;
+            &lt;span aria-hidden="true"&gt;&amp;times;&lt;/span&gt;&lt;/button&gt;
+          &lt;div id="${modalId}Label" class="modal-title"&gt;
```
Comment (trrenty) [inline xwiki-pro-macros-ui/src/main/resources/Confluence/Macros/ViewFile.xml:423]: escape modalId

File: xwiki-pro-macros-ui/src/main/resources/Confluence/Macros/ViewFile.xml [hunk 4]
```diff
@@ -263,24 +386,140 @@
...
+            $escapetool.xml($services.localization.render('rendering.macro.viewFiles.modal.title'))
+          &lt;/div&gt;
+        &lt;/div&gt;
+        &lt;div class="modal-body" style="max-height: 80vh; overflow: auto"&gt;
```
Comment (trrenty) [inline xwiki-pro-macros-ui/src/main/resources/Confluence/Macros/ViewFile.xml:427]: Why did you add the style here and not in a CSS extension?

File: xwiki-pro-macros-ui/src/main/resources/Confluence/Macros/ViewFile.xml [hunk 5]
```diff
@@ -263,24 +386,140 @@
...
+
+        #if ($modalId.contains('pdf'))
```
Comment (trrenty) [inline xwiki-pro-macros-ui/src/main/resources/Confluence/Macros/ViewFile.xml:429]: `$escapedFilename.endsWith(".pdf")` would be better instead of `contains`. Since we could have an office filename that contains "pdf".

File: xwiki-pro-macros-ui/src/main/resources/Confluence/Macros/ViewFile.xml [hunk 6]
```diff
@@ -263,24 +386,140 @@
...
+#end
+#macro (displayPresentationContent $escapedFilename, $modalId)
+  &lt;div class="modal fade" id="$modalId" tabindex="-1" role="dialog"
+      aria-labelledby="${modalId}Label"&gt;
+    &lt;div class="modal-dialog" style="height:90%; width:90%; display:flex; flex-direction:column; align-items:center;" role="document"&gt;
+      &lt;div class="modal-content"&gt;
+        &lt;div class="modal-header"&gt;
+          &lt;button type="button" class="close" data-dismiss="modal" aria-label="Close"&gt;
+            &lt;span aria-hidden="true"&gt;&amp;times;&lt;/span&gt;&lt;/button&gt;
+          &lt;div id="${modalId}Label" class="modal-title"&gt;
```
Comment (trrenty) [inline xwiki-pro-macros-ui/src/main/resources/Confluence/Macros/ViewFile.xml:452]: escape .modalId`

File: xwiki-pro-macros-ui/src/main/resources/XWiki/Macros/Translations.xml [hunk 7]
```diff
@@ -160,7 +160,12 @@ rendering.macro.userList.parameter.properties.description=List of user propertie
 rendering.macro.userList.parameter.fixedTableLayout.name=Fixed table layout
 rendering.macro.userList.parameter.fixedTableLayout.description=If the user table should have fixed column width across tables displaying the same user properties
 rendering.macro.userList.isEmpty=The User List macro has an empty user list.
-rendering.macro.userList.invalidUser=User {0} does not exist or is not viewable.</content>
+rendering.macro.userList.invalidUser=User {0} does not exist or is not viewable.
+
+## View Files macro
+rendering.macro.viewFiles.modal.title=Content view
```
Comment (trrenty) [inline xwiki-pro-macros-ui/src/main/resources/XWiki/Macros/Translations.xml:166]: you should also use translations for the macro parameters.

**Review by trrenty — CHANGES_REQUESTED** [review]

**Comment by trrenty** [comment]
I also think that it would be more intuitive to have something more stylized than a simple link. I think its not expected to have a link that opens a modal.

Maybe something like this:
image

Ofc this is subject to discussion. Maybe @snazare could give her opinion as well. And u, @ChiuchiuSorin , ofc :P

**Comment by ChiuchiuSorin** [comment]
Tried two thumbnail styles, I added for both of them a file type specific icon:
Style 1, using XWiki button wrapper:
style1
Style 2, using bootstrap default:
style2

Personally, I'd prefer the first option, as it is more visible and it adapts the aspect depending on the used skin
WDYT @snazare and @trrenty ?

**Comment by snazare** [comment]
Hi @ChiuchiuSorin I agree, let's go with the XWiki button so that it follows the look&feel defined in the instance.

File: xwiki-pro-macros-ui/src/main/resources/Confluence/Macros/ViewFile.xml [hunk 8]
```diff
@@ -263,24 +404,161 @@
...
+        #set ($modalId = $escapetool.xml($unescapedFilename.replaceAll("[^a-zA-Z]", "") + 'DisplayContentModal'))
+        #if ($presentationExtensions.contains($extension))
+          &lt;div class="buttonwrapper"&gt;
+            &lt;a href="#${modalId}" data-toggle="modal" data-target="#${modalId}"&gt;
+              $services.icon.render('file-powerpoint') $escapedFilename&lt;/a&gt;
+          &lt;/div&gt;
+          #displayPresentationContent($escapedFilename, $modalId)
+        #else
+          #if($extension == 'pdf')
+            &lt;div class="buttonwrapper"&gt;
```
Comment (trrenty) [inline xwiki-pro-macros-ui/src/main/resources/Confluence/Macros/ViewFile.xml:537]: You should also have a "title" attribute for the buttons that would say something along the lines "Click to view the file". To indicate further that the button is a thumbnail.

You shouldnt have an `a` tag inside the button. I reckon you placed it like this to have the `onhover` effect. It will probably trigger only when hovering the text inside the button. You should add a css rule for the button so that the text will get underlined even if you hover on the wrapper of the text.

You can also have a velocity macro for rendering the button, since you duplicate the code.

**Review by trrenty — CHANGES_REQUESTED** [review]

**Review by trrenty — APPROVED** [review]
